### ISSUE TITLE
conjugate heat transfer

### ISSUE THREAD

**Issue by sezerh24** [WaterLily-jl/WaterLily.jl#172]
Hi, 

I think this is an excellent code. 

is it possible to give an example of conjugate heat transfer for the flow around a cylinder to determine the Nusselt's number correlations? 

a similar example is given below: https://www.youtube.com/watch?v=skvU9fUaUzM

**Comment by b-fg** [comment]
We do not solve for temperature, but a temperature-transport equation could be included indeed. Still, this is not in our roadmap, but we would be happy to provide hints/support in doing so.

**Comment by sezerh24** [comment]
Thank you for your feedback. So, any scalar solver (i.e. Finite Volume Solver) can be integrated with waterlily.jl? I am using VoronoiFVM.jl(https://github.com/j-fu/VoronoiFVM.jl) and and JFVM.jl (https://github.com/simulkade/JFVM.jl). In VoronoiFVM.jl voronoi meshes are used, so it would be very difficult to map the velocity field (from waterlily) into the grid. However, JFVM is using cartessian grid. But it is not using immersed boundary method for curved geometries.

**Comment by b-fg** [comment]
No, I was not implying other solver can be integrated. I was mentioning that a scalar transport equation could be implemented in WaterLily, and solve it for temperature.

**Comment by sezerh24** [comment]
oh okay, thank you. So, how it could be done? I am not familiar with the waterlily architecture. would you be able to help me for the implementation of a scalar transport solver ?

**Comment by b-fg** [comment]
Not at the moment. But we can add it as a feature to be considered in our (now way to long) to-do list.

**Comment by sezerh24** [comment]
indeed, waterlily could be a very fast solver for reactive flow in complex vegetation and it can have great application to wildfires (As wildfire are in everywhere, now in California). Currently, FDS(FireDynamicsSimulator) form NIST is used for wildfires(WFDS) and fire simulations. However, the FDS is written in fortran and it is difficult to access the structure of the code, Also, FDS is not based on immersed boundary method. So, it is difficult to use it for complex geometries like vegetation. People used FDS to simulate the burning of Christmass tree, but it requires a long input files to complete the simulation and it is really challenging.. plus the simulated tree does not have the same geometry as the real tree. Also, FDS does not support GPU as far as I know. I think waterlily has a great potential for fire simulations. I am willing to contribute if needed. Thank you, Hayri

**Comment by weymouth** [comment]
I just did a Christmas tree 🎄 although it's not on fire and the geometry want very realistic 😉.

I think this would be a fun avenue to explore, but it's not my research direction, so someone else would need to lead. My suggestion would be to write a new extension to the `AbstractSimulation` class, similar to the BiotSavartBCc repository.

**Comment by tdunning** [comment]
@sezerh24 are you only thinking about the heat is being released as in the burning tree example, or is simple heat transfer as in a cool fluid flowing over warm structure like a radiator or heat sink interesting as well?

**Comment by sezerh24** [comment]
@tdunning I am thinking both... the object can generate heat... i.e. chemical reaction or reactive flow and or simply conjugate heat transfer like heat sink... so, we need to solve the scalar transport equation in general. 
Please the equation below... phi can be species concentration or temperature. 
<img width="524" alt="Image" src="https://github.com/user-attachments/assets/a93b3562-9a42-45cd-9f09-3da255f996d1" />

**Comment by tdunning** [comment]
That's really nice!

This would have to hook in with density as well if you want to see convection, I guess.

**Comment by TzuYaoHuang** [comment]
I recently start a side project on extending WaterLily.jl with Boussinesq approximation and BDIM on the temperature field. [This repo](https://github.com/TzuYaoHuang/BoussinesqApprox.jl) might be interesting for this issue.

**Comment by sezerh24** [comment]
Thank you, Huang. I did test your implementation, but unfortunately I wasn't able to reproduce the same results. It's very possible that I made a mistake in my setup or misunderstood part of the implementation, so I'll continue looking into it.

In parallel, I have developed a scalar transport solver on a Cartesian grid that is externally coupled with WaterLily.

<img width="731" height="596" alt="Image" src="https://github.com/user-attachments/assets/d7aa7298-219a-4e3d-92a4-4ccbdd499673" />

At the moment, my implementation does not directly integrate scalar transport into the WaterLily API. Instead, the flow solver and scalar solver are coupled externally. I have also implemented the Boussinesq approximation through forcing terms, and that approach has been working well.

The main drawback of my current framework is the external coupling. On the other hand, it gives me the flexibility to leverage the Julia ODE ecosystem for stiff chemical kinetics. This capability is particularly important for my research on pore-scale reactive porous media, where transport and reaction processes can be strongly coupled and highly stiff.

My code is not open source yet because it still needs significant cleanup and reorganization. I think your implementation could be very helpful as a reference. Eventually, I'd like to integrate my PDE solver more tightly with WaterLily, similar to your approach. My PDE framework already supports solving coupled systems of PDEs using both explicit and implicit time integration through the Julia ODE ecosystem, so a closer integration with WaterLily would be a natural next step.

My next goal is to extend WaterLily toward weakly compressible flow simulations for buoyancy-driven plumes and flame dynamics. I'm currently exploring possible approaches to achieve this while preserving WaterLily's efficiency. If you have any insights or suggestions on how you would approach weakly compressible flow within the WaterLily framework, I'd be very interested to hear your thoughts.

Thanks again for sharing your work!

**Comment by TzuYaoHuang** [comment]
Yeah, the code might be broken right now since it relies on a few WaterLily API tweaks I've proposed in a pending PR -- it should be working again once that gets merged!

Your simulation looks really interesting! Would you mind sharing the validation reference for your test case? I'd love to run a comparison on my end. I'd also be curious to hear how you're handling the immersed boundary condition for the temperature field, as heat transfer is relatively new territory for me.

AFAIK, WaterLily doesn't have plans to support implicit or higher-order time integration natively, so external coupling sounds like the way to leverage Julia's ODE ecosystem for stiff systems.

As for weakly compressible flow, it isn't really my area of expertise, but here is how I'd approach it architecturally: As now `Simulation` take custom `Flow` and `Poisson` structs, you could define a new compressible `Flow` wrapper struct containing the original `Flow` alongside fields like sound speed, then overload mom_step! (much like BiotSavartBCs.jl or my repo). In a similar fashion, you'd likely want a custom `Poisson` wrapper struct and a modified pressure propagation function to solve the updated pressure equation.

**Comment by sezerh24** [comment]
@TzuYaoHuang ,  thank you for your reply. The semester started for us, and I was very busy to reply.  Here's the validation reference for this case:

Setup: Re=500, Pr=0.71, blockage ratio = 0.1, isothermal cylinder, confined channel.

Cd = 1.35, Nu = 11.21 (time-averaged over the periodic vortex-shedding regime).

Cd is consistent with published confined-cylinder results at this Re and blockage. For Nu, I compared against a couple of sources: Rathore et al. (2006, Heat and Mass Transfer) report Nu ≈ 12.0–12.4 at the same β=0.1, Re=500, though at a slightly different Pr (0.744 vs my 0.71). Fu & Tong (2002, IJHMT) run a case at the same β=0.1 and Pr=0.71 as mine, and their Re=500 result shows Nu cycling roughly 10–11.8 with a cycle-average around 11.0–11.2, which lines up closely with my 11.21.

On the thermal IBM side: I don't do anything too elaborate. I solve the energy equation as-is over the whole domain, using the same BDIM velocity field that comes out of the flow solver. Since BDIM already forces velocity to zero inside the solid, the advection term u·∇T automatically vanishes there — no separate treatment needed for that part.

To actually enforce the constant-temperature (Dirichlet) condition on the cylinder wall, I use a penalty method: it's added as a source term directly in the energy equation, forcing T toward the prescribed wall temperature inside/at the solid region. So it's the same spirit as BDIM's velocity penalization, just applied to the temperature field instead — no ghost-cell reconstruction needed. It's a pretty direct approach and so far it's working well for me.

On weak compressibility — I'm genuinely very interested in this. Having weakly compressible capability in WaterLily would open the door to simulating fire plume dynamics, which is directly relevant to my research. The architectural approach you outlined (a compressible Flow wrapper struct with sound speed, overloading mom_step!, plus a modified Poisson wrapper for the updated pressure equation) sounds like the right direction to me, and I'd like to explore contributing to that if there's interest. Would be great to hear more of your thinking on how you'd handle the coupling between the compressible pressure equation and the existing projection step, since that's the part I'm least sure how to approach cleanly.
